README.md
# Colour

An easy way to use color in flutter (with Normal, HEX, RGB, ARGB, or RGBO format).

## Getting Started

Add dependency to your flutter project.

```
$ flutter pub add colour
```

or

```yaml
dependencies:
  colour: ^1.0.5
```

or from git repo:

```yaml
dependencies:
  colour:
    git: https://github.com/salkuadrat/colour.git
```

## Usage

You can use it like a normal Color:
```
Colour(0xFFFFFFFF)
Colour(0x00000000)

// with opacity
Colour(0xFFFFFFFF, 0.25)
Colour(0xFFFFFFFF, 0.5)
```

Or with any HEX color formats:
```
Colour('121212')
Colour('#121212')
Colour('af12dc')
Colour('#af12dc')
Colour('AF12DC')
Colour('#AF12DC')
Colour('FF121212')
Colour('#FF121212')
Colour('ff1212aa')
Colour('#ff1212aa')

// with opacity
Colour('121212', 0.25)
Colour('#121212', 0.5)
Colour('af12dc', 0.75)
Colour('#af12dc', 1.0)
Colour('AF12DC', 0.0)
Colour('#AF12DC', 0.25)
Colour('FF121212', 0.3)
Colour('#FF121212', 0.8)
Colour('ff1212aa', 0.6)
Colour('#ff1212aa', 0.42)
```

Feel free to use RGB format.
```
Colour(255, 255, 255)
Colour(0, 0, 0)
Colour(120, 120, 80)
```

It can also be used with ARGB format.
```
Colour(0, 255, 255, 255) // shorter than Color.fromARGB(0, 255, 255, 255)
Colour(255, 0, 0, 0)
Colour(75, 120, 120, 80)
```

RGBO format can works fine too...
```
Colour(255, 255, 255, 0.0) // shorter than Color.fromRGBO(255, 255, 255, 0.0)
Colour(0, 0, 0, 0.5)
Colour(120, 120, 80, 1.0)
```

Anyhow, this is a rare case. Suppose you already have a defined variable of Color.
You can apply opacity to Color using Colour like this:
```
// suppose you already have an existing Color
Color _color = Colors.teal;

// apply opacity with Colour 
Colour(_color, 0.75) // the same as _color.withOpacity(0.75)
```

# Visual Studio Code Extension

You can search and install extension named ```fluttercolour``` to show Colour preview.

![](assets/fluttercolour1.png)

![](assets/fluttercolour2.png)


CHANGELOG.md
## [1.0.5] - 19 May 2021

* Migrate to null safety

## [1.0.4] - 22 Sept 2020

* Add Visual Studio Code Extension

## [1.0.3] - 21 Sept 2020

* Fix formatting and add example.

## [1.0.2] - 21 Sept 2020

* Can accept instance of Color as the first parameter

## [1.0.1] - 21 Sept 2020

* Add a more restricted checking for RGB params.

## [1.0.0] - 21 Sept 2020

* A working Colour to use color in flutter in any of these formats: Normal, HEX, RGB, ARGB, or RGBO.


test/colour_test.dart
import 'package:flutter/material.dart';
import 'package:flutter_test/flutter_test.dart';

import 'package:colour/colour.dart';

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();

  test('Normal color', () {
    assert(Colour(0xFFFFFFFF).value == Color(0xFFFFFFFF).value);
    assert(Colour(0x00000000).value == Color(0x00000000).value);
  });

  test('Normal color with opacity', () {
    assert(Colour(0xFFFFFFFF, 0.25).value ==
        Color(0xFFFFFFFF).withOpacity(0.25).value);
    assert(Colour(0x00000000, 0.5).value ==
        Color(0x00000000).withOpacity(0.5).value);

    // use existing color with opacity
    Color color = Color(0xFFFFFFFF);
    assert(Colour(color, 0.75).value == color.withOpacity(0.75).value);
  });

  test('Normal hex color', () {
    assert(Colour('121212').value == Color(0xFF121212).value);
    assert(Colour('#121212').value == Color(0xFF121212).value);
  });

  test('Normal hex color with opacity', () {
    assert(Colour('121212', 0.25).value ==
        Color(0xFF121212).withOpacity(0.25).value);
    assert(Colour('#121212', 0.5).value ==
        Color(0xFF121212).withOpacity(0.5).value);
  });

  test('Hex color with transparency', () {
    assert(Colour('FF121212').value == Color(0xFF121212).value);
    assert(Colour('B1121212').value == Color(0xB1121212).value);
    assert(Colour('#B1121212').value == Color(0xB1121212).value);
    assert(Colour('#00121212').value == Color(0x00121212).value);
  });

  test('Hex color with transparency and opacity', () {
    assert(Colour('FF121212', 0.25).value ==
        Color(0xFF121212).withOpacity(0.25).value);
    assert(Colour('B1121212', 0.5).value ==
        Color(0xB1121212).withOpacity(0.5).value);
    assert(Colour('#B1121212', 0.25).value ==
        Color(0xB1121212).withOpacity(0.25).value);
    assert(Colour('#00121212', 0.5).value ==
        Color(0x00121212).withOpacity(0.5).value);
  });

  test('RGB color', () {
    assert(Colour(255, 255, 255).value ==
        Color.fromARGB(255, 255, 255, 255).value);
    assert(Colour(0, 0, 0).value == Color.fromARGB(255, 0, 0, 0).value);
    assert(Colour(123, 24, 80).value == Color.fromARGB(255, 123, 24, 80).value);
  });

  test('ARGB color', () {
    assert(Colour(255, 255, 255, 255).value ==
        Color.fromARGB(255, 255, 255, 255).value);
    assert(Colour(0, 0, 0, 0).value == Color.fromARGB(0, 0, 0, 0).value);
    assert(
        Colour(75, 123, 24, 80).value == Color.fromARGB(75, 123, 24, 80).value);
  });

  test('RGBO color', () {
    assert(Colour(255, 255, 255, 1.0).value ==
        Color.fromRGBO(255, 255, 255, 1.0).value);
    assert(Colour(0, 0, 0, 0.0).value == Color.fromRGBO(0, 0, 0, 0.0).value);
    assert(Colour(75, 123, 24, 0.25).value ==
        Color.fromRGBO(75, 123, 24, 0.25).value);
  });
}


example/lib/main.dart
import 'package:colour/colour.dart';
import 'package:flutter/material.dart';

void main() {
  runApp(MyApp());
}

class MyApp extends StatelessWidget {
  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      theme: ThemeData(
        primarySwatch: Colors.blue,
        visualDensity: VisualDensity.adaptivePlatformDensity,
      ),
      home: MyHomePage(title: 'Flutter Demo Home Page'),
    );
  }
}

class MyHomePage extends StatefulWidget {
  MyHomePage({Key? key, required this.title}) : super(key: key);

  final String title;

  @override
  _MyHomePageState createState() => _MyHomePageState();
}

class _MyHomePageState extends State<MyHomePage> {
  int _counter = 0;

  void _incrementCounter() {
    setState(() {
      _counter++;
    });
  }

  @override
  Widget build(BuildContext context) {
    return Scaffold(
      appBar: AppBar(
        title: Text(widget.title),
      ),
      body: Center(
        child: Column(
          mainAxisAlignment: MainAxisAlignment.center,
          children: <Widget>[
            Text(
              'You have pushed the button this many times:',
              style: TextStyle(
                color: Colour('424242'),
              ),
            ),
            Text(
              '$_counter',
              style: Theme.of(context).textTheme.headline4,
            ),
            Container(
              width: 72,
              height: 56,
              color: Colour('#38ef7d'),
            ),
            SizedBox(height: 10),
            Container(
              width: 72,
              height: 56,
              color: Colour('#f80759', 0.75),
            ),
          ],
        ),
      ),
      floatingActionButton: FloatingActionButton(
        onPressed: _incrementCounter,
        tooltip: 'Increment',
        child: Icon(Icons.add,
            color: Colour('efefef', 0.9)), // Custom color using Colour
      ), // This trailing comma makes auto-formatting nicer for build methods.
    );
  }
}


lib/colour.dart
library colour;

import 'package:flutter/material.dart';

/// An immutable 32 bit color value in any format (normal, hex, rgb, argb, rgbo).
///
///
/// Here are some ways it could be constructed:
///
/// ```dart
/// Colour c = const Colour(0xFF42A5F5);
/// Colour c = const Colour(0xFF42A5F5, 0.5);
/// Colour c = const Colour('42A5F5');
/// Colour c = const Colour('42A5F5', 0.5);
/// Colour c = const Colour('#42A5F5');
/// Colour c = const Colour('42a5f5');
/// Colour c = const Colour('#42a5f5', 0.8);
/// Colour c = const Colour(255, 255, 255); // RGB
/// Colour c = const Colour(255, 255, 255, 255); // ARGB
/// Colour c = const Colour(255, 255, 255, 0.5); // RGBO
/// ```
///
class Colour extends Color {
  Colour(dynamic color, [dynamic x, dynamic y, dynamic z])
      : super(getColor(color, x, y, z));

  /// get an immutable 32 bit color value from dynamic parameters
  static int getColor(dynamic color, [dynamic x, dynamic y, dynamic z]) {
    bool hasX = x != null;
    bool hasY = y != null;
    bool hasZ = z != null;

    bool noX = x == null;
    bool noY = y == null;
    bool noZ = z == null;

    // check if Colour only use the fist paramter (color)
    // it could be a normal color: Colour(0xFF121212)
    // or a hex based color: Colour('#FF121212') or Colour('FF121212')
    // or Colour('#121212') or Colour('121212')
    bool isSingleParams = noX && noY && noZ;

    if (isSingleParams) {
      if (color is Color) return color.value;
      if (color is int) return color; // Normal color: Colour(0xFF121212)
      return getColorFromHex(color.toString());
    }

    // check if Colour use the first and second parameter (color and x)
    // it can apply to any of 5 formats:
    // Colour(0xFF121212, 0.25)
    // Colour('#FF121212', 0.25)
    // Colour('FF121212', 0.25)
    // Colour('#121212', 0.25)
    // Colour('121212', 0.25)
    bool isDoubleParams = hasX && noY && noZ;
    bool isXOpacity = x is double && x >= 0.0 && x <= 1.0;

    if (isDoubleParams && isXOpacity) {
      Color _color;
      if (color is Color)
        _color = color;
      else if (color is int)
        _color = Color(color);
      else
        _color = Color(getColorFromHex(color.toString()));
      return _color.withOpacity(x).value;
    }

    // check if Colour use only 3 parameters (color, x and y)
    // then it is RGB color => Colour(255, 255, 255)
    bool isTripleParams = hasX && hasY && noZ;
    bool isTripleParamsInteger = color is int && x is int && y is int;

    if (isTripleParams && isTripleParamsInteger) {
      return Color.fromARGB(255, color, x, y).value;
    }

    // check if Colour use all 4 parameters
    // it can apply to any of 2 formats
    // ARGB => Colour(255, 12, 123, 255)
    // RGBO => Colour(12, 123, 255, 0.25)
    bool isAllParams = hasX && hasY && hasZ;

    if (isAllParams) {
      // if all parameters is int, then it is ARGB => Colour(255, 12, 123, 255)
      bool isARGB = color is int && x is int && y is int && z is int;

      if (isARGB) {
        return Color.fromARGB(color, x, y, z).value;
      }

      // if the last paramter is double, then it is RGBO => Colour(12, 123, 255, 0.25)
      bool isRGBO = color is int &&
          x is int &&
          y is int &&
          z is double &&
          z >= 0.0 &&
          z <= 1.0;

      if (isRGBO) {
        return Color.fromRGBO(color, x, y, z).value;
      }
    }

    return color;
  }

  /// parsing hex color for any of the four formats, such as:
  /// getColorFromHex('#FF121212')
  /// getColorFromHex('FF121212')
  /// getColorFromHex('#121212')
  /// getColorFromHex('121212')
  static int getColorFromHex(String color) {
    String hex = color.toUpperCase().replaceAll('#', '');
    if (hex.length == 6) hex = 'FF' + hex;
    return int.parse(hex, radix: 16);
  }
}
